### ISSUE TITLE
Host crashes when client joins after U148

### ISSUE THREAD

**Issue by IKillerwolfI** [polarathene/biglobby#50]
Played with some mates but when the update came we crashed the game of our current host.
But when I hosted one friend was able to join but not the other 4 (nobody crashed this time).
We all had the same version and could play pre-update.

Application has crashed: access violation

-------------------------------

Callstack:

         payday2_win32_release  (???)     ???                                                 
         payday2_win32_release  (???)     ???                                                 
         payday2_win32_release  (???)     ???                                                 

-------------------------------

Current thread: Main
Script stack:
       orig_NetworkPeer_send()  lib/network/base/networkpeer.lua:1034               
                        send()  @mods/BeardLib/Hooks/NetworkPeer.lua:64             
                        send()  @mods/biglobby/lua/lib/network/base/_networkpeer.lua:14  
    on_join_request_received()  @mods/biglobby/lua/lib/network/base/session_states/_hoststateinlobby.lua:117  
    on_join_request_received()  lib/network/base/hostnetworksession.lua:55          
                                lib/network/base/handlers/connectionnetworkhandler.lua:50  

-------------------------------

**Comment by polarathene** [comment]
See my response [here](https://github.com/polarathene/biglobby/issues/48#issuecomment-307813036). There is no network changes in the usual file, this might not be an update I can allocate unpaid time to fix any time soon :(

If the func_name that is causing the crash does not require too much time to investigate and fix I'll hopefully be able to get biglobby working for you again, else it will require someone from the community to contribute.

**Comment by miaulightouch** [comment]
uh, here is no issue with U148, though I only install big lobby and custom translation blt mod.
and it's work properly when playing with 7 real players in 16 players room.

consider to update your BLT and reinstall your mods?

**Comment by IKillerwolfI** [comment]
Well you only got BLT but we also have BeardLib installed which has something to do with the crash:


Current thread: Main
Script stack:
       orig_NetworkPeer_send()  lib/network/base/networkpeer.lua:1034               
                        send()  @mods/**BeardLib**/Hooks/NetworkPeer.lua:64             <-----
                        send()  @mods/biglobby/lua/lib/network/base/_networkpeer.lua:14  
    on_join_request_received()  @mods/biglobby/lua/lib/network/base/session_states/_hoststateinlobby.lua:117  
    on_join_request_received()  lib/network/base/hostnetworksession.lua:55          
                                lib/network/base/handlers/connectionnetworkhandler.lua:50

**Comment by polarathene** [comment]
@IKillerwolfI And there is no crash if you disable biglobby, BeardLib causes no problem?

**Comment by KarmaPanda** [comment]
I tried with only biglobby and it still crashed.

**Comment by IKillerwolfI** [comment]
I tried it without biglobbby and every other mod still on and it didn't crash.
So yes the game doesn't crash when BeardLib is active but biglobby is disabled.

**Comment by polarathene** [comment]
These line looks like the only one from my mod:
```
send() /lua/lib/network/base/_networkpeer.lua:14
on_join_request_received() /lua/lib/network/base/session_states/_hoststateinlobby.lua:117
```

BeardLib gets a network request from biglobby and has trouble handling it:
```
orig_NetworkPeer_send() lib/network/base/networkpeer.lua:1034
send() @mods/BeardLib/Hooks/NetworkPeer.lua:64
```
I'd need to know what that network request is. It's possible that BeardLib remove my biglobby network methods and it doesn't understand what they are when it gets them for some reason. You might be able to fix this by adjusting the mod.txt priority value(if one is higher than the other than just swap them around I guess?).

I don't have time to look into it too much atm. I haven't had issues reported in the past about BeardLib being a problem so maybe it's a recent change they made? If so maybe that mod has the bug not biglobby?

**Comment by Luffyyy** [comment]
As it seems the requests both mods handle are 'set_unit' and 'join_request_reply', it crashes after it continues to the normal function.
in set_unit we basically change a parameter(which is 4 in NetworkPeer:send) if it contains any custom weapons/misc.
in join_request_reply we change the 15th parameter(if the current map is custom) to a fixed one that contains the custom map, I'd assume you aren't using a custom map, they can be a little broken in the current public version.

BeardLib doesn't change func_name at all.

**Comment by polarathene** [comment]
> BeardLib doesn't change func_name at all.

Clearly, based on your description it's modifying how a method is expected to work. It's doing what biglobby did in alpha before requiring a pdmod.

[I do send an extra parameter](https://github.com/polarathene/biglobby/blob/master/biglobby/lua/lib/network/base/session_states/_hoststateingame.lua#L118) for `join_request_reply`, `on_join_request_received()` isn't a clean method to modify, but it's signature is untouched. `on_join_request_reply()`([handler](https://github.com/polarathene/biglobby/blob/master/biglobby/lua/lib/network/base/handlers/_connectionnetworkhandler.lua#L12)) [supports native and modded versions](https://github.com/polarathene/biglobby/blob/master/biglobby/lua/lib/network/base/_clientnetworksession.lua#L5), it detects if the original params exist or if there is an extra param and that it is a number. Other than that I guess if someone tries similar it'll break. 

I think I could prefix `on_join_request_reply()` with `BigLobby__` as I've done with my own connection and unit network handlers and methods in network_settings. Not sure if that'd cause any issues with my mod or break others from working as expected like BeardLib expecting joins to happen like normal.

BigLobby doesn't do anything with `set_unit` afaik besides raise peer limit. Any methods altered by biglobby in the network_settings are all prefixed and handled by their own handler calling the same functions. You're changing what is sent in the 4th parameter ["outfit version"](https://github.com/polarathene/biglobby/blob/d4dde691d9beb579b344ba1f74f6f33bd02b4f26/pdmod%20src/mod/network.network_settings#L387)? Only reason I can see that causing a problem is if you did some logic in the set_unit unit network handler before calling it's method. If that is the case, might be fixable if biglobby currently is applied first, if were applied after BeardLib it should inherit any modifications to unit_network_handler. If this is where you handle `join_request_reply`, there would be a problem.

---

Just for clarity, when a network message is to be sent and it matches one of those biglobby modifies, it gets a prefix prepended to that func_name. That could maybe be a cause for breakage too with other mods.

**Comment by IKillerwolfI** [comment]
So will you be able to fix it this weekend? 
Or is there something I can do that would temporarily fix it?

**Comment by polarathene** [comment]
Uhh the fix is use either BeardLib or BigLobby, not both :P

---

BeardLib is altering how two methods params are expected to be from what @Luffyyy said. While they could possibly use the same approach I do now, it would conflict with my mod as only one mod is able to modify network_settings file with pdmod. 

I need confirmation what the conflict causing the crash is. I don't think it's `set_unit` but it could be. Sounds like `join_request_reply` is where it goes wrong, but where specifically I don't know. Sounds like either It's BeardLibs version getting normal params but expecting their modification when it's not, or it's my version getting the modified param and not making sense of it.

Any mod could do the same thing that BeardLib does, so I'm not adding code to my method to cater for modified params like they've done. I can include a network method in my pdmod specifically for BeardLib compatibility but this requires BeardLib to do the same or at least detect BigLobby in use and use that version instead. **I doubt they're keen on special detection as well. If they're ok with using pdmod for a network method, we could agree on a network method that works for any mod wanting to pass extra param to this method.**

---

A fix might be to have some generic network methods for passing data around,  having `join_request_reply` packing params for mods in a string and unpacking when received, as a new param, not modifying an existing one for that method. I'm not sure if either of these are solutions, as depending what is happening it might not be easy for both our mods to send the data we want and receive it. I'd be willing to give it some more thought once I know what the current flow is that is causing the breakage, I do not have time to investigate this right now as I've not done pd2 mod dev for some time.

**Comment by IKillerwolfI** [comment]
I found out that the issue was a mod called VELOsync. 
We fixed the game by deleting said mod and re downloading biglobby so Beardlib wasn't the problem in the first place. 
Anyways we only found this out by removing all mods except beardlib and biglobby and it worked. 
Then we added all mods back except the ones were i though they might cause the crash (VELOsync was one of them). 
So I'm closing this here thanks for your reply and I hope we won't run in problems like this in the future again.
